source/extensions/filters/network/dubbo_proxy/buffer_helper.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace DubboProxy {
+
+/**
+ * BufferWrapper provides a partial implementation of Buffer::Instance that is sufficient for
+ * BufferHelper to read protocol data without draining the buffer's contents.
+ */
+class BufferWrapper : public Buffer::Instance {
```

Comment:
I feel like we have this code in a bunch of different places now. Is it different or the same? I think we have variants here, Mongo, Thrift, it was in the MySQL PR but I'm not sure that ever got followed up on (I can find it).

I'm just wondering if we can consolidate this since it's generally useful? This is going to come up when working on Kafka also. 

@zuercher thoughts here? 

cc @adamkotwasinski  for Kafka

Reply:
The only thing I'd be concerned here is avoiding potential buffer underflow.

In case of large structure that's split across multiple `ReadFilter::onData` invocations (== multiple `Buffer` instances, we might need to keep a copy of some part of currently received data.

In case of array or arrays (with array serialized as `length` + `N * element`), for large payloads we might something have the content split across multiple calls - what means that no matter how we implement that, we are going to need a buffer.
(see that with redis' filter implementation, where `DecoderImpl::decode` does keep state, and it looks like it can handle large payloads (== multiple `onData` calls).

Making envoy call filter's `onData` with small (degenerate even - single byte) buffers might show some interesting results. Guess I'll need to cover that in my Kafka tests.